src/util/sort_key.rs

```diff
@@ -0,0 +1,85 @@
+// Rust Bitcoin Library
+// Written in 2020 by
+//     rust-bitcoin developers
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the CC0 Public Domain Dedication
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//
+
+//! Utility for key Sorting
+//!
+//! Sort `bitcoin::PublicKey` according to the logic implemented in
+//! bitcoin core. In bitcoin core, keys are merely wrappers over
+//! bytes and the comparison function directly checks the underlying
+//! bytes.
+//! When keys are compressed, this utility can be used for BIP67
+//! comparison.
+use std::cmp;
+
+use PublicKey;
+
+/// Utility structure for representing ordering of PublicKeys
+/// as implemented in bitcoin core. Compares the underlying
+/// bytes when comparing the two PublicKeys in compressed form
+/// instead of naturally derived order which compares underlying
+/// 64 byte secp key.
+#[derive(Debug, Eq, PartialEq, Clone)]
+pub struct SortKey<'pk>(&'pk PublicKey);
+
+impl<'pk> From<&'pk PublicKey> for SortKey<'pk> {
+    fn from(pk: &'pk PublicKey) -> Self {
+        SortKey(pk)
+    }
+}
+
+impl<'pk> cmp::Ord for SortKey<'pk> {
+    fn cmp(&self, other: &Self) -> cmp::Ordering {
+        match (self.0.compressed, other.0.compressed) {
+            (true, false) => cmp::Ordering::Less,
+            (false, true) => cmp::Ordering::Greater,
+            (true, true) => self.0.key.serialize().cmp(&other.0.key.serialize()),
+            (false, false) => self
+                .0
+                .key
+                .serialize_uncompressed()
+                .cmp(&other.0.key.serialize_uncompressed()),
+        }
+    }
+}
+
+impl<'pk> PartialOrd for SortKey<'pk> {
```

Comment:
BTW, according to BIP-67, "Uncompressed keys are incompatible with this specificiation. A compatible implementation should not automatically compress keys. Receiving an uncompressed key from a multisig participant should be interpreted as a sign that the user has an incompatible implementation."

So my proposal here is to make `PartialOrd` returning `None` if any of the keys is uncompressed (and leave `Ord` as is)

Reply:
Or, alternatively, two types should be used: `Bip67Key`, which will be `PartialOrd` only; and `SortedKey` for Bitcoin Core-style sorting with the current code unchainged. Top-namespace `cmp` function should be split into two, `bip67_partial_cmp -> Option<cmp::Ordering>` and `cmp`, left as is.

Reply:
I like the second idea with two types. In fact, my primary motivation was to have only the `SortKey` type to emulate core behnaviour with no focus on `bip67` all. This is because in `sortedmulti()` descriptor, bitcoin core does allow uncompressed keys and sorts them with this logic. 

I will amend the PR with two separate types.